### PR TITLE
Add messages, catalog, parse namespaces

### PR DESCRIPTION
**Pull request by nick-mccoy**
# Description of change
Adds 3 new namespaces
- messages
- catalog
- parse

# Manual QA steps
 - wrote unit tests for all funcs
 
# Risks
 - n/a
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: test/singer_clojure/catalog_test.clj [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                   "stream_c" { "stream"       "stream_c"
+                               "metadata"      {"selected" true}}
+                   "stream_d" { "stream"       "stream_d"
+                               "metadata"      {"selected" false}}
+                   "stream_e" { "stream"       "stream_e"
+                               "metadata"      {"selected" true}}}]
+      (is (= (get-selected-streams catalog)
+             '("stream_a" "stream_c" "stream_e"))))))
+
+(def test-serialized-catalog {"streams"
```
Comment (asaf-erlich) [inline test/singer_clojure/catalog_test.clj]: Would it benefit to add a comment here that this is used in other tests as well?

File: test/singer_clojure/messages_test.clj [hunk 2]
```diff
@@ -0,0 +1,136 @@
+(ns singer-clojure.messages-test
+  (:require [singer-clojure.messages :as singer-messages]
+            [clojure.test :refer [is deftest]]
+            [clojure.string :as string]
+            [clojure.data.json :as json]))
+
+(defmacro with-out-str-and-value
```
Comment (asaf-erlich) [inline test/singer_clojure/messages_test.clj:7]: Should we add a docstring here to explain the use case / why we added this macro? That we wanted both the stdout AND the returned value.

File: test/singer_clojure/messages_test.clj [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  [& body]
+  `(let [s# (new java.io.StringWriter)]
+     (binding [*out* s#]
+       (let [v# ~@body]
+         (vector (str s#)
+                 v#)))))
+
+(defn get-messages-from-output
```
Comment (asaf-erlich) [inline test/singer_clojure/messages_test.clj:16]: Similar docstring might be useful

File: src/singer_clojure/catalog.clj [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+(defn get-selected-streams
+  [catalog]
+  "Takes a deserialized catalog and returns a sequence of the selected
+  tap-stream-ids"
+  (filter #(get-in catalog [% "metadata" "selected"]) (keys catalog)))
+
+;;
+;;
+;; SERIALIZATION
```
Comment (asaf-erlich) [inline src/singer_clojure/catalog.clj]: Is this comment necessary?

File: src/singer_clojure/catalog.clj [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  [catalog]
+  (-> catalog
+      serialize-streams
+      serialize-metadata))
+
+
+
+;;
+;;
+;; DESERIALIZATION
```
Comment (asaf-erlich) [inline src/singer_clojure/catalog.clj]: Same here?

File: src/singer_clojure/messages.clj [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+(defn write!
+  "Writes a valid singer message to stdout"
+  [message]
+  {:pre [(is (valid? message) (format "message %s is not valid" message))]}
+  (-> message
+      (json/write-str)
+      println))
+
+(defn make-unsupported-schemas-empty [schema-message catalog stream-name]
```
Comment (dmosorast) [inline src/singer_clojure/messages.clj]: Implementation details like this could probably be marked private (`defn-`)

File: src/singer_clojure/catalog.clj [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  (update stream-catalog-entry "schema"
+          serialize-stream-schema))
+
+(defn serialize-streams
+  [catalog]
+  (assoc {}
+         "streams"
+         (map serialize-stream (vals catalog))))
+
+(defn serialize-catalog
```
Comment (KAllan357) [inline src/singer_clojure/catalog.clj:59]: Is it the case that this function and the opposing deserialize are the only real "public" functions in this namespace? I think noting the other functions with `defn-` would be useful in denoting that they are "private" and not really part of the API that should be used.

File: src/singer_clojure/messages.clj [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+         "RECORD"
+         (message "record")
+
+         "ACTIVATE_VERSION"
+         (message "version"))))
+
+(defn write!
+  "Writes a valid singer message to stdout"
+  [message]
+  {:pre [(is (valid? message) (format "message %s is not valid" message))]}
```
Comment (luandy64) [inline src/singer_clojure/messages.clj:25]: Do we need this `is`? If so, then should we be consistent and make the other pre check use `is` and have a meaningful message?

File: test/singer_clojure/catalog_test.clj [hunk 9]
```diff
@@ -0,0 +1,93 @@
+(ns singer-clojure.catalog-test
```
Comment (dmosorast) [inline test/singer_clojure/catalog_test.clj:1]: I think this looks familiar, just wanted to comment that there should be a bunch of tests for this in tap-mssql, if you didn't copy this from there, that might be a gold mine of conditions.

File: src/singer_clojure/messages.clj (hunk 8 above)
Comment (luandy64) [reply src/singer_clojure/messages.clj:25]: The other place being `singer_clojure.messages/deserialize-stream`

**Review by dmosorast — CHANGES_REQUESTED** [review]
Seems pretty solid, just one comment and a potential change.

File: src/singer_clojure/messages.clj (hunk 8 above)
Comment (luandy64) [reply src/singer_clojure/messages.clj:25]: Maybe we need to make `singer_clojure.messages/deserialize-stream` because `singer_clojure.parse/state` also uses `is`

File: src/singer_clojure/parse.clj [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+(def cli-options
+  [["-d" "--discover" "Discovery Mode"]
+   [nil "--repl" "REPL Mode"]
+   [nil "--config CONFIG" "Config File"
+    :parse-fn #'config]
+   [nil "--catalog CATALOG" "Singer Catalog File"
+    :parse-fn (comp catalog/deserialize-catalog catalog)]
+   [nil "--state STATE" "Singer State File"
+    :parse-fn #'state]
```
Comment (luandy64) [inline src/singer_clojure/parse.clj]: Should these all get short flags too? It would be consistent with `singer-python`

File: test/singer_clojure/catalog_test.clj (hunk 1 above)
Comment (nick-mccoy) [reply test/singer_clojure/catalog_test.clj]: Fixed

File: test/singer_clojure/messages_test.clj (hunk 2 above)
Comment (nick-mccoy) [reply test/singer_clojure/messages_test.clj:7]: Fixed

File: test/singer_clojure/messages_test.clj (hunk 3 above)
Comment (nick-mccoy) [reply test/singer_clojure/messages_test.clj:16]: Fixed

File: src/singer_clojure/catalog.clj (hunk 4 above)
Comment (nick-mccoy) [reply src/singer_clojure/catalog.clj]: Removed

File: src/singer_clojure/catalog.clj (hunk 5 above)
Comment (nick-mccoy) [reply src/singer_clojure/catalog.clj]: Removed

File: src/singer_clojure/catalog.clj (hunk 7 above)
Comment (nick-mccoy) [reply src/singer_clojure/catalog.clj:59]: Yeah, I forgot about marking private functions with `defn-` -- These are updated now

File: src/singer_clojure/messages.clj (hunk 6 above)
Comment (nick-mccoy) [reply src/singer_clojure/messages.clj]: Fixed

File: src/singer_clojure/parse.clj (hunk 10 above)
Comment (nick-mccoy) [reply src/singer_clojure/parse.clj]: Good call -- added the short flag case as well

File: src/singer_clojure/messages.clj (hunk 8 above)
Comment (asaf-erlich) [reply src/singer_clojure/messages.clj:25]: Using is here was a way to create a message when the condition failed. There was a stackoverflow post that suggested this: https://stackoverflow.com/questions/17390962/can-i-include-a-message-in-pre-and-postconditions

But I agree that being consistent and using this for all pre/post conditions should probably be done. We only added it when we saw one failed but it provided no information to debug why it failed.

File: test/singer_clojure/catalog_test.clj (hunk 9 above)
Comment (asaf-erlich) [reply test/singer_clojure/catalog_test.clj:1]: Yeah we'll take a look. Obviously we took code from the existing taps and sometimes tests too. We'll take one more look there in case.

File: test/singer_clojure/catalog_test.clj [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+                   "stream_d" { "stream"       "stream_d"
+                               "metadata"      {"selected" false}}
+                   "stream_e" { "stream"       "stream_e"
+                               "metadata"      {"selected" true}}}]
+      (is (= (get-selected-streams catalog)
+             '("stream_a" "stream_c" "stream_e"))))))
+
+;; `test-serialized-catalog` and `test-deserialized-catalog` are def'd
+;; here since they are also used in the parse tests
+(def serialized-test-catalog {"streams"
```
Comment (asaf-erlich) [inline test/singer_clojure/catalog_test.clj:23]: You renamed the method but left the comment referring to the old method name. Maybe just remove the name from the comment?

File: src/singer_clojure/messages.clj [hunk 12]
```diff
@@ -0,0 +1,83 @@
+(ns singer-clojure.messages
+  (:require [clojure.data.json :as json]
+            [clojure.test :refer [is]]))
+
+(defn valid?
```
Comment (asaf-erlich) [inline src/singer_clojure/messages.clj:5]: Should this be defn- (private)? I actually don't know. It's only used as the precondition in the write! method.

File: src/singer_clojure/parse.clj [hunk 13]
```diff
@@ -0,0 +1,47 @@
+(ns singer-clojure.parse
+  (:require [clojure.data.json :as json]
+            [clojure.tools.cli :as cli]
+            [clojure.java.io :as io]
+            [clojure.test :refer [is]]
+            [singer-clojure.catalog :as catalog]))
+
+(defn slurp-json
```
Comment (asaf-erlich) [inline src/singer_clojure/parse.clj:8]: Same question here, should this be defn-?

File: test/singer_clojure/catalog_test.clj (hunk 11 above)
Comment (asaf-erlich) [reply test/singer_clojure/catalog_test.clj:23]: Just went ahead and fixed the names.

File: src/singer_clojure/messages.clj (hunk 8 above)
Comment (asaf-erlich) [reply src/singer_clojure/messages.clj:25]: Fixed, will update PR shortly

File: test/singer_clojure/catalog_test.clj (hunk 9 above)
Comment (asaf-erlich) [reply test/singer_clojure/catalog_test.clj:1]: So all those tests are creating and destroying databases, which is specific to ms sql and not necessarily the library code. We looked through them to see if there was an edge case we missed but we believe the test we have is end-to-end enough to cover what we want to have tested.

Comment (nick-mccoy) [reply test/singer_clojure/catalog_test.clj:1]: We were checking that out and most of them weren't generic enough to put into the library, so we decided to just write a super generic serialize/deserialize test

File: src/singer_clojure/messages.clj (hunk 12 above)
Comment (asaf-erlich) [reply src/singer_clojure/messages.clj:5]: We decided to leave these as is

File: src/singer_clojure/parse.clj (hunk 13 above)
Comment (asaf-erlich) [reply src/singer_clojure/parse.clj:8]: We decided to leave these as is

File: test/singer_clojure/catalog_test.clj (hunk 9 above)
Comment (dmosorast) [reply test/singer_clojure/catalog_test.clj:1]: Cool, sounds good

File: src/singer_clojure/parse.clj [hunk 14]
```diff
@@ -23,7 +23,7 @@
 (defn state
   "This function exists as a test seam and for the post condition"
   [state-file]
-  {:post [(is (map? %) (format "map %s is not valid" %))]}
+  {:post [(is (map? %) (format "state-file %s is not a map" %))]}
```
Comment (dmosorast) [inline src/singer_clojure/parse.clj]: Might be good to use the name here. I didn't even realize the `%` version of the parameter would work here.

**Review by dmosorast — CHANGES_REQUESTED** [review]
One more question/comment/change

File: src/singer_clojure/parse.clj (hunk 14 above)
Comment (asaf-erlich) [reply src/singer_clojure/parse.clj]: Sure

Comment (asaf-erlich) [reply src/singer_clojure/parse.clj]: So % is the return value so we're trying to figure out how to refer to that in a named way, because the message is wrong, it's not that state-file is a map, but slurp-json of state-file is

Comment (dmosorast) [reply src/singer_clojure/parse.clj]: Ahhh, okay, that makes sense. :+1:

**Review by dmosorast — APPROVED** [review]
